File: src/mjorbit/cpu/orbit/lvlh.py
"""LVLH frame construction and transforms.

LVLH convention (RSW / Hill frame):
  x̂ = radial (R̂ = R/|R|)
  ŷ = along-track (h × R̂ direction, approximately velocity for circular)
  ẑ = cross-track (h = R × V direction)

Wait — standard RSW:
  R̂ = radial outward
  Ŝ = in orbital plane, perpendicular to R̂, in direction of motion
  Ŵ = orbit normal (R × V)

We use RSW = LVLH:
  x̂_L = R̂
  ŷ_L = Ŝ (along-track in orbital plane)
  ẑ_L = Ŵ = (R × V) / |R × V|

Units: km, km/s, rad/s.
"""

from __future__ import annotations

import numpy as np
from mjorbit.cpu.orbit.state import OrbitState, FrameCache
from mjorbit.cpu.orbit.gravity import total_accel


def update_frame_cache(orbit: OrbitState, use_j2: bool = False) -> FrameCache:
    """Build LVLH frame from chief orbit state.

    Args:
        orbit: current chief orbit state
        use_j2: include J2 in omega_dot computation (uses finite differences)

    Returns:
        FrameCache with rotation matrices, omega, omega_dot
    """
    R, V = orbit.R_eci, orbit.V_eci
    r = np.linalg.norm(R)

    # Basis vectors
    x_hat = R / r  # radial
    h = np.cross(R, V)
    h_mag = np.linalg.norm(h)
    z_hat = h / h_mag  # cross-track (orbit normal)
    y_hat = np.cross(z_hat, x_hat)  # along-track

    # C_LI: rows are LVLH basis vectors expressed in ECI
    C_LI = np.stack([x_hat, y_hat, z_hat], axis=0)  # shape (3,3)
    C_IL = C_LI.T

    # Angular velocity of LVLH frame: omega = h / r^2 along orbit normal
    # In LVLH coordinates: omega = (0, 0, h/r^2) but let's express in ECI first
    omega_eci = h / r**2  # = (R x V) / r^2
    omega_lvlh = C_LI @ omega_eci  # express in LVLH

    # omega_dot: d/dt(omega) in LVLH frame
    # omega_eci = h / r^2, where h = R x V
    # d(omega)/dt = dh/dt / r^2 - 2 * h * rdot / r^3
    # For point-mass: dh/dt = R x a = R x (-mu/r^3 R) = 0
    # For J2: dh/dt = R x a_J2 != 0 in general
    a_eci = total_accel(R, use_j2=use_j2)
    dh_dt = np.cross(R, a_eci)
    dr_dt = np.dot(R, V) / r  # d|R|/dt
    domega_dt_eci = dh_dt / r**2 - 2.0 * h * dr_dt / r**3
    omega_dot_lvlh = C_LI @ domega_dt_eci

    return FrameCache(
        C_LI=C_LI,
        C_IL=C_IL,
        omega_lvlh=omega_lvlh,
        omega_dot_lvlh=omega_dot_lvlh,
    )


def eci_to_lvlh_pos(r_eci: np.ndarray, R_ref: np.ndarray, C_LI: np.ndarray) -> np.ndarray:
    """Convert ECI position to LVLH position relative to chief.

    Args:
        r_eci: absolute ECI position, km
        R_ref: chief ECI position, km
        C_LI: LVLH-from-ECI rotation

    Returns:
        relative position in LVLH, km
    """
    return C_LI @ (r_eci - R_ref)


def lvlh_to_eci_pos(r_lvlh: np.ndarray, R_ref: np.ndarray, C_IL: np.ndarray) -> np.ndarray:
    """Convert LVLH relative position to absolute ECI position."""
    return R_ref + C_IL @ r_lvlh


def eci_to_lvlh_vel(
    v_eci: np.ndarray,
    r_lvlh: np.ndarray,
    V_ref: np.ndarray,
    C_LI: np.ndarray,
    omega_lvlh: np.ndarray,
) -> np.ndarray:
    """Convert ECI velocity to LVLH relative velocity.

    v_lvlh = C_LI @ (v_eci - V_ref) - omega x r_lvlh
    """
    return C_LI @ (v_eci - V_ref) - np.cross(omega_lvlh, r_lvlh)

File: src/cpp/src/lvlh.cc
#include "mjorbit/lvlh.h"

#include "mjorbit/gravity.h"
#include "mjorbit/math_utils.h"

namespace mjorbit {

void update_frame_cache(
    const double R_eci[3],
    const double V_eci[3],
    FrameCache* out_cache,
    bool use_j2) {
  update_frame_cache(R_eci, V_eci, out_cache, use_j2, CentralBodySpecNative{});
}

void update_frame_cache(
    const double R_eci[3],
    const double V_eci[3],
    FrameCache* out_cache,
    bool use_j2,
    const CentralBodySpecNative& central_body) {
  if (!out_cache) {
    return;
  }

  const double r = detail::norm3(R_eci);
  double h[3];
  detail::cross3(R_eci, V_eci, h);
  const double h_mag = detail::norm3(h);
  if (r == 0.0 || h_mag == 0.0) {
    detail::set_identity3(out_cache->C_LI);
    detail::set_identity3(out_cache->C_IL);
    detail::zero3(out_cache->omega_lvlh);
    detail::zero3(out_cache->omega_dot_lvlh);
    return;
  }

  double x_hat[3];
  detail::normalize3(R_eci, x_hat);
  double z_hat[3];
  detail::normalize3(h, z_hat);
  double y_hat[3];
  detail::cross3(z_hat, x_hat, y_hat);

  out_cache->C_LI[0] = x_hat[0];
  out_cache->C_LI[1] = x_hat[1];
  out_cache->C_LI[2] = x_hat[2];
  out_cache->C_LI[3] = y_hat[0];
  out_cache->C_LI[4] = y_hat[1];
  out_cache->C_LI[5] = y_hat[2];
  out_cache->C_LI[6] = z_hat[0];
  out_cache->C_LI[7] = z_hat[1];
  out_cache->C_LI[8] = z_hat[2];
  detail::mat3_transpose(out_cache->C_LI, out_cache->C_IL);

  double omega_eci[3];
  detail::scale3(h, 1.0 / (r * r), omega_eci);
  detail::mat3_mul_vec(out_cache->C_LI, omega_eci, out_cache->omega_lvlh);

  double a_eci[3];
  total_accel(R_eci, a_eci, use_j2, central_body);
  double dh_dt[3];
  detail::cross3(R_eci, a_eci, dh_dt);
  const double dr_dt = detail::dot3(R_eci, V_eci) / r;
  double domega_dt_eci[3];
  for (int i = 0; i < 3; ++i) {
    domega_dt_eci[i] = dh_dt[i] / (r * r) - 2.0 * h[i] * dr_dt / (r * r * r);
  }
  detail::mat3_mul_vec(out_cache->C_LI, domega_dt_eci, out_cache->omega_dot_lvlh);
}

void eci_to_lvlh_pos(const double r_eci[3], const double R_ref[3], const double C_LI[9], double out_r_lvlh[3]) {
  double delta[3];
  detail::sub3(r_eci, R_ref, delta);
  detail::mat3_mul_vec(C_LI, delta, out_r_lvlh);
}

void lvlh_to_eci_pos(const double r_lvlh[3], const double R_ref[3], const double C_IL[9], double out_r_eci[3]) {
  double delta_eci[3];
  detail::mat3_mul_vec(C_IL, r_lvlh, delta_eci);
  detail::add3(R_ref, delta_eci, out_r_eci);
}

void eci_to_lvlh_vel(
    const double v_eci[3],
    const double r_lvlh[3],
    const double V_ref[3],
    const double C_LI[9],
    const double omega_lvlh[3],
    double out_v_lvlh[3]) {
  double delta_v[3];
  detail::sub3(v_eci, V_ref, delta_v);
  double delta_v_lvlh[3];
  detail::mat3_mul_vec(C_LI, delta_v, delta_v_lvlh);
  double omega_cross_r[3];
  detail::cross3(omega_lvlh, r_lvlh, omega_cross_r);
  detail::sub3(delta_v_lvlh, omega_cross_r, out_v_lvlh);
}

}  // namespace mjorbit
